### DISCUSSION TITLE
importlib.metadata.PackageNotFoundError: regex while building an .exe with pyinstaller

### DISCUSSION THREAD

**Discussion by alinacoding** [pyinstaller/pyinstaller#6033] [Help] [answered]
I am trying to build an executable with pyinstaller and I am getting the following errors:

Traceback (most recent call last):
  File "transformers/utils/versions.py", line 105, in require_version
  File "importlib/metadata.py", line 551, in version
  File "importlib/metadata.py", line 524, in distribution
  File "importlib/metadata.py", line 187, in from_name
importlib.metadata.PackageNotFoundError: regex

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "chatbot2.py", line 8, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 546, in exec_module
  File "transformers/__init__.py", line 43, in <module>
  File "PyInstaller/loader/pyimod03_importers.py", line 546, in exec_module
  File "transformers/dependency_versions_check.py", line 41, in <module>
  File "transformers/utils/versions.py", line 120, in require_version_core
  File "transformers/utils/versions.py", line 107, in require_version
importlib.metadata.PackageNotFoundError: The 'regex!=2019.12.17' distribution was not found and is required by this application. 
Try: pip install transformers -U or pip install -e '.[dev]' if you're working with git master
[30391] Failed to execute script 'chatbot2' due to unhandled exception!

I am using PyInstaller version 5.0.dev0 and Python version 3.9.1.

I tried pip3.9 install transformers, pip3.9 install regex, I tried adding hooks hook-regex.py and hook-transformers.py. I would really appreciate any suggestions! Thank you!

**Comment by rokm** [comment, marked as answer]
For every package listed in [`pkgs_to_check_at_runtime`](https://github.com/huggingface/transformers/blob/534f6eb9f1f8910a4912ccccd79f1f974c625168/src/transformers/dependency_versions_check.py#L26) (except `python`), you will need to collect metadata, by using `copy_metadata(name)` in your spec file.

 For quick prototyping, you can also use `--copy-metadata <name>` command-line switch (i.e., add `--copy-metadata tqdm --copy-metadata regex --copy-metadata sacremoses ...` to your PyInstaller command-line).

And I seem to recall that at least one of those packages also requires a `--collect-data` (or `collect_data_files()` in spec).

**Reply by alinacoding** [reply]
Thank you so much for your very quick answer! It worked! I've now moved on to another error!

**Reply by wzhappapp** [reply]
Can I see your sepc? Thanks a lot!

**Reply by wzhappapp** [reply]
I've got it! Thanks!

**Comment by seanbenhur** [comment]
Where should we include copy_metatsata in spec file?

**Reply by rokm** [reply]
Somewhere before the `Analysis` is created. You need to use it to populate a list that you pass to the `Analysis` via its `datas=` argument.

You can use `--copy-metadata` command-line switch and look at the resulting .spec file.

**Comment by Litchilitchy** [comment]
@rokm Thanks for your hints, I use `--copy-metadata` and get a `.spec` like this
```python
from PyInstaller.utils.hooks import copy_metadata

datas = []
datas += copy_metadata('tqdm')

block_cipher = None

a = Analysis(
    ['scripts\\my_script.py'],
    pathex=['.'],
    binaries=[],
    datas=datas,
```
I then run `pyinstaller my_script.spec` and got the executable. But when I run the .exe it still got 
```
Traceback (most recent call last):
  File "transformers\utils\versions.py", line 108, in require_version
  File "importlib\metadata.py", line 569, in version
  File "importlib\metadata.py", line 542, in distribution
  File "importlib\metadata.py", line 196, in from_name
importlib.metadata.PackageNotFoundError: tqdm

During handling of the above exception, another exception occurred:
...
File "transformers\utils\versions.py", line 110, in require_version
importlib.metadata.PackageNotFoundError: The 'tqdm>=4.27' distribution was not found and is required by this application.
```

**Reply by rokm** [reply]
Hmm, works for me. Can you try with `pyinstaller --clean my_script.spec`?

**Reply by Litchilitchy** [reply]
Great, it works! It would be better if this part is in official document.

**Comment by duhoang** [comment]
I had the same problem and adding the meta data fixed it. But then I get this other error:

`  File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
  File "torch\_dynamo\skipfiles.py", line 86, in <module>
    ] + [
        ^
  File "torch\_dynamo\skipfiles.py", line 88, in <listcomp>
    _module_dir(m)
  File "torch\_dynamo\skipfiles.py", line 76, in _module_dir
    return _strip_init_py(m.__file__)
                          ^^^^^^^^^^
AttributeError: module 'abc' has no attribute '__file__'. Did you mean: '__name__'?
[10852] Failed to execute script 'main' due to unhandled exception!`

Is this because I didn't collect_data_files correctly?

**Reply by rokm** [reply]
Hmm, no. This looks like a bug in PyInstaller; with python >= 3.11, `abc` is a frozen module, but it seems to have `__file__` attribute when imported in unfrozen python. Whereas when imported in PyInstaller-frozen script, it doesn't have the `__file__` attribute, hence the error.

**Reply by duhoang** [reply]
Is there a workaround or should I go back to previous Python versions?

**Reply by rokm** [reply]
I'll have to dig around the python's loader code to see where the difference is coming from, so for the time being, you're better going back to older python version.

**Reply by rokm** [reply]
Or patch `torch\_dynamo\skipfiles.py` to gracefully handle the lack of `__file__` attribute on modules.
